core/shared/src/main/scala/io/chrisdavenport/log4cats/LoggerFactory.scala

```diff
@@ -0,0 +1,13 @@
+package io.chrisdavenport.log4cats
+
+trait LoggerFactory[F[_]] {
```

Comment:
We also need methods that do not return in a logger in `F[_]`. Probably in a separate trait, in which we express like stuff [getLogger: SelfAwareStructuredLogger](https://github.com/ChristopherDavenport/log4cats/blob/master/slf4j/src/main/scala/io/chrisdavenport/log4cats/slf4j/Slf4jLogger.scala#L29).

And your `Slf4jLoggerFactory` would implement both.

Reply:
+1 for adding impure but pragmatic `getXxx(…): Logger[F]` methods.

I disagree with adding them in a separate trait though. The biggest virtue of `LoggerFactory[F]` is that the whole application and libraries as well may depend on a common factory to spawn their loggers. And if there would be two or more traits, one would have to make a tough choice for the whole app which trait to use or, alternatively, pass both factories bloating the code.

Reply:
The reason I opt for a separate trait is that sometimes it is impossible to write a factory that returns things in non-F in a safe way. For instance loggers that require some internal buffer that then send stuff somewhere remote — like to AWS cloudwatch or something. You'd need to instantiate some mutable state inside an `F[_]` upon logger creation.

What we can do, maybe, is add a kitchen-sync included factory that inherits from both? Although this might cause the diamond problem in implicit resolution. But if your codebase does not have such stateful loggers, then you can just use the super-trait everywhere and no harm done. But if you do have statefull loggers, then you cannot use the super-trait everywhere.

Reply:
> loggers that require some internal buffer that then send stuff somewhere remote

That's an interesting point. To put this idea even further perhaps `F[Logger[F]]` would not be enough here and `Resource[F, Logger[F]]` would be more appropriate.

I wonder if it's possible to move the state and other heavy machinery insider the factory itself? We would then have something like this
```scala
object CloudwatchLoggerFactory {
  def resource[F: …](…): Resource[F, LoggerFactory[F]] = …
}
``` 
while its `getLogger`s would just set a logger name and thus could be implemented trivially.

_Disclaimer: I'm a bit inexperienced with remote logging, so all of the above might make no sense._

Reply:
I implemented [something for Cloudwatch logging](https://github.com/busymachines/pureharm-aws/blob/master/aws-logger/src/main/scala/busymachines/pureharm/aws/logger/AWSLoggerFactory.scala), and currently only the `LoggerFactory` creation itself has to be a `Resource`. I didn't get to implement the local buffering yet, that's why my factory has non-F methods, but I think that a simple `F[_]` is enough.

Reply:
But that doesn't mean that `Resource[F *]` won't be necessary in other cases 😄

Reply:
@lavrov, https://gitter.im/ChristopherDavenport/libraries?at=5e5413111ec15e02119ae192